helper/schema/resource_diff.go

```diff
@@ -0,0 +1,453 @@
+package schema
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/terraform/terraform"
+)
+
+// newValueWriter is a minor re-implementation of MapFieldWriter to include
+// keys that should be marked as computed, to represent the new part of a
+// pseudo-diff.
+type newValueWriter struct {
+	*MapFieldWriter
+
+	// A list of keys that should be marked as computed.
+	computedKeys map[string]bool
+
+	// A lock to prevent races on writes. The underlying writer will have one as
+	// well - this is for computed keys.
+	lock sync.Mutex
+}
+
+// WriteField overrides MapValueWriter's WriteField, adding the ability to flag
+// the address as computed.
+func (w *newValueWriter) WriteField(address []string, value interface{}, computed bool) error {
+	if err := w.MapFieldWriter.WriteField(address, value); err != nil {
+		return err
+	}
+
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	if w.computedKeys == nil {
+		w.computedKeys = make(map[string]bool)
+	}
+
+	if computed {
+		w.computedKeys[strings.Join(address, ".")] = true
```

Comment:
To be entirely honest I'm not sure what the necessary behavior is here... eventually this all gets flattened to strings in the current diff format, so I would guess that we'd want to set this to something that eventually becomes the empty string, but I might be wrong about that. If you're seeing Terraform currently generating diffs where e.g. computed ints appear as `"0"` in the diff then what you did here _is_ probably the right behavior, even though it feels a little counter-intuitive.